contracts/epochs/FlowEpoch.cdc

```diff
@@ -109,46 +121,511 @@ pub contract FlowEpochs {
+                }
+            }
+        }
+
+        pub fun endStakingAuction() {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only end staking auction during the staking auction"
+            }
+
+            FlowEpoch.endStakingAuction()
+
+            FlowEpoch.startEpochSetup(randomSource: unsafeRandom().toString())
+        }
+
+        pub fun endEpochSetup() {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.EPOCHSETUP: "Can only end Epoch Setup during epoch setup"
+            }
+            FlowEpoch.endEpochSetup()
+
+            FlowEpoch.startEpochCommitted()
+        }
+
+        pub fun endEpoch() {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.EPOCHCOMMITTED: "Can only end epoch during Epoch Committed"
+            }
+
+            FlowEpoch.startNewEpoch()
+        }
+    }
+
+    /// Pays rewards, moves staking tokens between buckets,
+    /// and starts the new epoch staking auction
+    access(account) fun startNewEpoch() {
+
+        self.stakingAdmin.payRewards()
+
+        self.stakingAdmin.moveTokens()
+
+        self.currentEpochPhase = EpochPhase.STAKINGAUCTION
+
+        // Update the epoch counters
+        self.currentEpochCounter = self.proposedEpochCounter
+        self.proposedEpochCounter = self.currentEpochCounter + 1 as UInt64
+    }
+
+    /// Ends the staking Auction with all the proposed nodes approved
+    access(account) fun endStakingAuction() {
+        let ids = FlowIDTableStaking.getProposedNodeIDs()
+
+        let approvedIDs: {String: Bool} = {}
+        for id in ids {
+            // Here is where we would make sure that each node's 
+            // keys and addresses are correct, they haven't committed any violations,
+            // and are operating properly
+            // for now we just set approved to true for all
+            approvedIDs[id] = true
+        }
+
+        self.stakingAdmin.endStakingAuction(approvedNodeIDs: approvedIDs)
+    }
+
+    /// Starts the EpochSetup phase and emits the epoch setup event
+    access(account) fun startEpochSetup(randomSource: String) {
+
+        // Get all the nodes that are proposed for the next epoch
+        let ids = FlowIDTableStaking.getProposedNodeIDs()
+
+        // Holds the node Information of all the approved nodes
+        var nodeInfoArray: [FlowIDTableStaking.NodeInfo] = []
+
+        // Holds node IDs of only collector nodes for QC
+        var collectorNodeIDs: [String] = []
+
+        // Holds node IDs of only consensus nodes for DKG
+        var consensusNodeIDs: [String] = []
+
+        // Get nodeinfo for all the nodes
+        // get all the collector and consensus nodes
+        // to initialize the QC and DKG
+        for id in ids {
+            let nodeInfo = FlowIDTableStaking.NodeInfo(nodeID: id)
+
+            nodeInfoArray.append(nodeInfo)
+
+            if nodeInfo.role == 1 as UInt8 {
```

Comment:
Have you considered using enums for the role and not just a number? Or at least assign the number as a contant and use that as a variable here. For me this flags as 'magical number'

Reply:
The roles are defined in the staking contract, which was written before enums we possible in Cadence, and upgrading state in the contract is very unsafe, so were stuck with `UInt8` for now. Do you think it would be useful to define an enum in this contract for it and just convert for everything in here?

Reply:
I think so yes, it makes the code here easier to understand. And it is already a handful :)

Reply:
ok I'll do that! 👍